app/src/lib/git/stash.ts

```diff
@@ -144,23 +144,62 @@ function extractBranchFromMessage(message: string): string | null {
   return match === null || match[1].length === 0 ? null : match[1]
 }
 
-/** Get the files that were changed in the given stash commit.
- *  This is different than `getChangedFiles` because stashes
- *  have _3 parents(!!!)_
+/**
+ * The SHA for the [empty tree](https://stackoverflow.com/questions/9765453/is-gits-semi-secret-empty-tree-object-reliable-and-why-is-there-not-a-symbolic).
+ *
+ */
+const NullTreeSHA = '4b825dc642cb6eb9a060e54bf8d69288fbee4904'
+
+/**
+ * Get the files that were changed in the given stash commit.
  *
- * TODO: support showing untracked files in the diff
+ * Runs `git diff` twice, once for tracked changes in the stash
+ * and once for untracked changes in the stash.
+ * This is different than `getChangedFiles` because stashes
+ * have _3 parents(!!!)_
  */
 export async function getStashedFiles(
   repository: Repository,
-  sha: string
+  committish: string
 ): Promise<ReadonlyArray<CommittedFileChange>> {
   // opt-in for rename detection (-M) and copies detection (-C)
   // this is equivalent to the user configuring 'diff.renames' to 'copies'
   // NOTE: order here matters - doing -M before -C means copies aren't detected
+  const baseArgs = ['diff', '-C', '-M', '--name-status', '-z']
+  const trackedArgs = [...baseArgs, committish, `${committish}^`, '--']
+  const trackedResult = await git(
+    trackedArgs,
+    repository.path,
+    'getStashedFiles (tracked)'
+  )
+
+  const trackedFiles = parseChangedFiles(trackedResult.stdout, committish)
 
-  // git diff -C -M --name-status -z [STASH_SHA] [STASH_SHA]^
-  const args = ['diff', '-C', '-M', '--name-status', '-z', sha, `${sha}^`, '--']
-  const result = await git(args, repository.path, 'getStashedFiles')
+  const untrackedArgs = [...baseArgs, NullTreeSHA, `${committish}^3`, '--']
+  const untrackedResult = await git(
+    untrackedArgs,
+    repository.path,
+    'getStashedFiles (untracked)',
+    {
+      // if this fails, its most likely
+      // because there weren't any untracked files,
+      // and that's okay!
+      successExitCodes: new Set([0, 128]),
+    }
+  )
+
+  // if that command is successful and has output, we'll parse it
+  // and merge it with the other output
+  if (untrackedResult.exitCode === 0 && untrackedResult.stdout.length > 0) {
+    const untrackedFiles = parseChangedFiles(
+      untrackedResult.stdout,
+      `${committish}^3`
+    )
+
+    // order is important here, since we want untracked changes to
+    // override potential (and uncommon) collisions with tracked changes
+    return [...trackedFiles, ...untrackedFiles]
```

Comment:
Is this actually going to address the collisions? I worry that doing an object spread of two arrays is equivalent to concatenating them, which means the potential for duplicate values?

I put together this test to validate my assumption, and it currently fails:

```diff
diff --git a/app/test/unit/git/stash-test.ts b/app/test/unit/git/stash-test.ts
index b42581923..4b7f6a9b0 100644
--- a/app/test/unit/git/stash-test.ts
+++ b/app/test/unit/git/stash-test.ts
@@ -10,6 +10,7 @@ import {
   getLastDesktopStashEntryForBranch,
   dropDesktopStashEntry,
   popStashEntry,
+  getStashedFiles,
 } from '../../../src/lib/git/stash'
 import { getStatusOrThrow } from '../../helpers/status'
 import { AppFileStatusKind } from '../../../src/models/status'
@@ -249,6 +250,43 @@ describe('git/stash', () => {
       expect(files).toHaveLength(1)
     })
   })
+
+  describe('getStashedFiles', () => {
+    let repository: Repository
+    let readme: string
+
+    beforeEach(async () => {
+      repository = await setupEmptyRepository()
+      readme = path.join(repository.path, 'README.md')
+      await FSE.writeFile(readme, 'text before removing from index')
+      await GitProcess.exec(['add', 'README.md'], repository.path)
+      await GitProcess.exec(['commit', '-m', 'initial commit'], repository.path)
+    })
+
+    it('displays uncommitted change rather than index change', async () => {
+      // stage the delete in the index
+      await GitProcess.exec(['rm', 'README.md'], repository.path)
+
+      // add a new file to the working directory (should be untracked)
+      await FSE.writeFile(readme, 'new untracked files contents')
+
+      // stash both changes
+      await stash(repository, 'master', null)
+
+      const entries = await getDesktopStashEntries(repository)
+      expect(entries.length).toBe(1)
+
+      const { stashSha } = entries[0]
+
+      const files = await getStashedFiles(repository, stashSha)
+
+      console.log('found files', files)
+
+      expect(files.length).toBe(1)
+
+      expect(files[0].status.kind).toBe(AppFileStatusKind.Untracked)
+    })
+  })
 })
 
 /**
@@ -264,7 +302,13 @@ async function stash(
   message: string | null
 ): Promise<void> {
   const result = await GitProcess.exec(
-    ['stash', 'push', '-m', message || createDesktopStashMessage(branchName)],
+    [
+      'stash',
+      'push',
+      '--include-untracked',
+      '-m',
+      message || createDesktopStashMessage(branchName),
+    ],
     repository.path
   )
```

But its not failing quite in the way I was expecting:

```

  ● git/stash › getStashedFiles › displays uncommitted change rather than index change

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 2

      283 |       console.log('found files', files)
      284 |
    > 285 |       expect(files.length).toBe(1)
          |                            ^
      286 |
      287 |       expect(files[0].status.kind).toBe(AppFileStatusKind.Untracked)
      288 |     })

      at Object.it (test/unit/git/stash-test.ts:285:28)

  console.log app/test/unit/git/stash-test.ts:283
    found files [ CommittedFileChange {
        path: 'README.md',
        status: { kind: 'Modified' },
        id: 'Modified+README.md',
        commitish: '8cebe5d30b27e96f7d3941579fd0d0dc78af7c83' },
      CommittedFileChange {
        path: 'README.md',
        status: { kind: 'New' },
        id: 'New+README.md',
        commitish: '8cebe5d30b27e96f7d3941579fd0d0dc78af7c83^3' } ]
```

And what's interesting is that, when we try and pop this stash, we encounter an error because `README.md` already exists:

```
$ git stash pop
README.md already exists, no checkout
Could not restore untracked files from stash entry
```

So I'm not too worried about this change given how tricky it is to setup, but I still think implementing way to filter out duplicate paths like how we do in `buildStatusMap` here would be helpful:

https://github.com/desktop/desktop/blob/4d6037998083199ae29a1a633afb286abc1b6ec7/app/src/lib/git/status.ts#L254-258

Reply:
ack, you are totally right, i spaced on that. fixing now!

Reply:
@shiftkey how does https://github.com/desktop/desktop/pull/7316/commits/6371695b2a6854a06367a1b8dfd9032b7b6ee332 feel? we're not using maps here so i went with something more array-ish